### DISCUSSION TITLE
Full-page lazy loading - custom scripts not working

### DISCUSSION THREAD

**Discussion by francoism90** [livewire/livewire#6749] [1. Help]
See:
https://livewire.laravel.com/docs/lazy#full-page-lazy-loading

This is my blade:
```
<button @click="foo">

@push('scripts')
    <script>
        function foo(e) {
            console.log(e)
        }
    </script>
@endpush
```
If I disable `lazy()` loading on the route, it works fine.

Is this normal behavior? I've tried injecting the script global, but this doesn't work either.

Thanks!

**Comment by PhiloNL** [comment]
Thanks for reporting. I've submitted a failing test for this issue https://github.com/livewire/livewire/pull/6757.

**Reply by francoism90** [reply]
@PhiloNL Thanks!

Yeah, I can confirm Alpine's x-init seems to be working, however it doesn't seem to re-init script tags.

If I should opt for a better solution, please let me know. :)

**Reply by PhiloNL** [reply]
I personally almost never use `<script>`. I have everything inside my `app.js` and this is the only file loaded. I use Alpine to initiate things instead of `<script>` tags as you can do `<div x-data x-init="callMyFunction"></div>` for example. But I do believe this is a bug.

**Reply by francoism90** [reply]
@PhiloNL It's weird, because I tried adding the script to `app.js`.

The only downside is the `app.js` size, but Vite solves this by using chunks I believe.

**Comment by Marroco93** [comment]
Is there any update in this? Im trying to use alpine init but still showing the components all over around and then show the skeleton loader

**Comment by jscott-rawnet** [comment]
I too would be keen to hear of any progress. I'm unsure moving js to a app.js would be suitable. The whole point surely is to have self-contained components. For example if i want to use alpine in my livewire component i want to be able to define my data from inside the component blade and pass properties from laravel:

```
<div x-data="data()">
    <ul class="mb-4 gap-y-1.5 flex flex-col" >
        <template x-for="(resource, index) in slice" :key="resource.id">
            <li><a :href="resource.url" class="underline" x-html="resource.name"></a><i
                    class="las la-angle-right"></i></li>
        </template>
    </ul>
    <button @click="showMore()" x-show="visibleResources < resources.length"
            class="rounded-md bg-brand-gray-400 text-white px-3 py-1 text-sm flex items-center justify-center gap-x-1">
        <i class="las la-plus text-white text-xl"></i>
        <span x-text="hiddenItemCount + ' More resources'"></span>
    </button>
    <button @click="showLess()" x-show="visibleResources > visibleResourcesDefault"
            class="rounded-md bg-brand-gray-400 text-white px-3 py-1 text-sm flex items-center justify-center gap-x-1">
        <i class="las la-minus text-white text-xl"></i>
        Less resources
    </button>
</div>
<script>
    function data() {
        return {
            visibleResourcesDefault: {!! $visibleResourcesDefault !!},
            visibleResources: {!! $visibleResourcesDefault !!},
            resources: {!! $resources !!},
            get slice() {
                return this.resources.slice(0, this.visibleResources);
            },
            get hiddenItemCount() {
                return this.resources.length - this.visibleResources;
            },
            showMore() {
                this.visibleResources = this.resources.length;
            },
            showLess() {
                this.visibleResources = this.visibleResourcesDefault;
            }
        }
    }
</script>
```

**Reply by potsky** [reply]
Same here. Migrating from v2 to v3, `<script>` injection in blade does not work anymore.

In my case, I can have several maps on the same page and using Leaflet, I have to defined dynamic functions like this:

```html
<div>
  <script type="text/javascript">
    var leaflet{{ $mapId }} = () => {
      var markersData = {!! json_encode($markers) !!};
      ...
    });

    leaflet{{ $mapId }}();
```

Trick here works fine for me : https://github.com/livewire/livewire/pull/6757#issuecomment-1804079723

**Comment by ysalitrynskyi** [comment]
This bug still exists.

**Comment by Marroco93** [comment]
The best solution for now is just put x-data="DataFunction".  and in my app.js and use this function Alpine.data('DataFunction', () => ({
//blablabla
}));
For now is what I can do to work with alpine and livewire

**Comment by Jamiul-Bari** [comment]
Still cannot use custom script in lazy loaded components. Not even inside @script @endscript

v3.4.10

**Comment by danie-ramdhani** [comment]
works fine for me....

```blade
<div>
    <div x-data="someAlpineComponent">
        <span x-text="console.log(foo)"></span>
    </div>
    
    <div x-init="someFunction"></div>
</div>

@script
<script>
    Alpine.data('someAlpineComponent', () => ({
        foo: 'bar',
    }));

    window.someFunction = () => {
        console.log('foo');
    }
</script>
@endscript
```

**Reply by Jamiul-Bari** [reply]
My bad. Full page lazy loading custom script works. I was testing components (not full-page) inside a view.

**Reply by Jamiul-Bari** [reply]
Steps to replicate:

1. Fresh Laravel project with livewire/livewire
2. `npm i swiper`
3. in `app.js`
```
import './bootstrap';

import Swiper from "swiper/bundle";
import "swiper/css/bundle";

window.Swiper = Swiper;
```
4. in `routes/web.php`
```
Route::get('x', fn() => view('x'));
```

5.  make component
```
php artisan make:livewire banner
```

6. `x.blde.php`
```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    @vite(['resources/js/app.js'])
</head>
<body>
    <livewire:banner lazy />
</body>
</html>
```

7. `livewire/banner.blade.php`
```
<div>
    <!-- Swiper -->
    <div class="swiper mySwiper">
        <div class="swiper-wrapper">
            <div class="swiper-slide">Slide 1</div>
            <div class="swiper-slide">Slide 2</div>
            <div class="swiper-slide">Slide 3</div>
            <div class="swiper-slide">Slide 4</div>
            <div class="swiper-slide">Slide 5</div>
            <div class="swiper-slide">Slide 6</div>
            <div class="swiper-slide">Slide 7</div>
            <div class="swiper-slide">Slide 8</div>
            <div class="swiper-slide">Slide 9</div>
        </div>
        <div class="swiper-button-next"></div>
        <div class="swiper-button-prev"></div>
        <div class="swiper-pagination"></div>
    </div>
</div>

@script
    <script>
        new Swiper(".mySwiper", {
            spaceBetween: 30,
            centeredSlides: true,
            autoplay: {
                delay: 500,
                disableOnInteraction: false,
            },
            pagination: {
                el: ".swiper-pagination",
                clickable: true,
            },
            navigation: {
                nextEl: ".swiper-button-next",
                prevEl: ".swiper-button-prev",
            },
        });
    </script>
@endscript
```

Now if you go to `http://laravel.test/x` and check if wont work. But in Step 6, if you remove `lazy` it will work
```
<livewire:banner />
```

[sample code reference](https://codesandbox.io/p/sandbox/kdsdk3?file=%2Findex.html%3A74%2C1-92%2C1)

**Reply by Kenneth-Inkum** [reply]
Same problem as mine.

**Reply by francoism90** [reply]
Could you try the same, but use x-intersect?

I believe you need to wrap this, and use the unit() method, as seen in the Alpinejs docs.

**Reply by vincentdev15** [reply]
Is it normal that the bug still exists while the bug has been reported quite one year ago ?

**Reply by lrljoe** [reply]
The "bug" was remediated here:
https://github.com/livewire/livewire/releases/tag/v3.2.0

**Reply by francoism90** [reply]
@wifirevaithi Just to be sure, did you add a placeholder? And you may need to add wire:ignore a level higher/use a different key.

I do try to use a simple blade for selects thought, it works a bit better when combining this with a form.

**Comment by Sohail9004** [comment]
I'm also facing same issue after adding the lazy loading but below solution is worked for me

I've added this code in the mount function of the component class

`
        $this->js(<<<JS
            const script = document.createElement('script');
            script.src = '/path_of_script/script.js';
            script.defer = true;
            document.body.appendChild(script);
        JS);
`

**Comment by i-begashvili** [comment]
my Temporary solution before comes Bug fix update (i guess..)

instead of lazy function, im using old practice to make as lazy..  $readyToLoad property with calling from blade makeReady() function in livewire by wire:init

```
<div wire:init="makeReady">
    @if ($readyToLoad)
        <div class="swiper tire-request-swiper">
            <div class="swiper-wrapper">
                @foreach ($requests as $item)
                    ...
                @endforeach
            </div>

            <!-- Swiper navigation buttons -->
            <div class="swiper-button-prev"></div>
            <div class="swiper-button-next"></div>
        </div>
    @endif


    <script>
        //script logic as function
        function initSwiper() {
            // Important to use timeout
            setTimeout(function() {
                new Swiper(".tire-request-swiper", {
                    autoplay: {
                        delay: 5000,
                    },
                    navigation: {
                        nextEl: ".swiper-button-next",
                        prevEl: ".swiper-button-prev",
                    },
                });
            });
        };
        
        // then we need this hook on success and call func
        Livewire.hook('commit', ({ 
            succeed,
        }) => {
            succeed(() => {
                initSwiper();
            });
        });
    </script>
</div>
```
So we need js timeout inside function and then call using Livewire hook 
than it works.
